### PR TITLE
Added HDF support for cases without a second input (target)

### PR DESCRIPTION
**Pull request by ArneNx**
Some small modifications to `HDFDataset` and `dump_from_dataset` to make it work with data that does not have a target key. 
So this works e.g. if you use `HDFDataset` on a `LMDataset` with delayed sequence disabled. 
I need this for masked language modeling and others might find this useful as well.

### CODE REVIEW

File: HDFDataset.py [hunk 1]
```diff
@@ -965,9 +965,13 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
     print("Data keys:", data_keys, file=log.v3)
     if "orth" in data_keys:  # special workaround for now, not handled
       data_keys.remove("orth")
-    data_target_keys = [key for key in dataset.get_target_list() if key in data_keys]
+
+    if len(data_keys) > 1:
```
Comment (albertz) [inline HDFDataset.py]: Why is this needed?
Also, isn't this wrong?

File: HDFDataset.py [hunk 2]
```diff
@@ -1039,15 +1047,18 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
         progress_bar_with_time(float(i) / num_seqs)
 
     print("Set seq len info...", file=log.v3)
-    hdf_dataset.create_dataset(attr_seqLengths, shape=(num_seqs, 2), dtype="int32")
+    hdf_dataset.create_dataset(attr_seqLengths, shape=(num_seqs, 1+len(data_target_keys)), dtype="int32")
```
Comment (albertz) [inline HDFDataset.py]: Please write PEP conform code (spacing etc; use PyCharm or any proper editor).

File: HDFDataset.py [hunk 3]
```diff
@@ -1039,15 +1047,18 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
-      if targets_len is None:
-        targets_len = data_len
-      hdf_dataset[attr_seqLengths][i] = [data_len, targets_len]
+      if len(data_target_keys) > 0:
+        targets_len = seq_len[default_data_target_key]
+        for data_key in data_target_keys:
+          assert seq_len[data_key] == targets_len, "different lengths in multi-target not supported"
+        if targets_len is None:
+          targets_len = data_len
+        hdf_dataset[attr_seqLengths][i] = [data_len, targets_len]
```
Comment (albertz) [inline HDFDataset.py]: This code is wrong now. E.g. if `len(data_target_keys) == 2`, then you must set 3 lengths here.

File: HDFDataset.py [hunk 4]
```diff
@@ -1039,15 +1047,18 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
-      targets_len = seq_len[default_data_target_key]
-      for data_key in data_target_keys:
-        assert seq_len[data_key] == targets_len, "different lengths in multi-target not supported"
-      if targets_len is None:
-        targets_len = data_len
-      hdf_dataset[attr_seqLengths][i] = [data_len, targets_len]
+      if len(data_target_keys) > 0:
+        targets_len = seq_len[default_data_target_key]
+        for data_key in data_target_keys:
+          assert seq_len[data_key] == targets_len, "different lengths in multi-target not supported"
```
Comment (albertz) [inline HDFDataset.py]: This is actually not needed anymore with your change of the `attr_seqLengths` shape above.

File: HDFDataset.py [hunk 5]
```diff
@@ -1039,15 +1047,18 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
+        if targets_len is None:
+          targets_len = data_len
+        hdf_dataset[attr_seqLengths][i] = [data_len, targets_len]
+      else:
+        hdf_dataset[attr_seqLengths][i] = [data_len]
```
Comment (albertz) [inline HDFDataset.py]: Do you actually need the `else` case? If you fix the code, probably it should directly work in all cases.

File: HDFDataset.py (hunk 1 above)
Comment (ArneNx) [reply HDFDataset.py]: I don't see why this should be wrong.
If you only have one data_key, you want it to go into data_input_keys, isn't that right?

File: HDFDataset.py (hunk 3 above)
Comment (ArneNx) [reply HDFDataset.py]: That's interesting. I did not intend to do anything here except enclosing it into an if block. I think something got messed up in a merge. I'll change it back.

File: HDFDataset.py (hunk 1 above)
Comment (albertz) [reply HDFDataset.py]: Ah, sorry, you are right.

File: HDFDataset.py [hunk 6]
```diff
@@ -976,14 +980,18 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
+        else:
+          raise Exception("not sure which target data key to use from %r" % (data_target_keys,))
       else:
-        raise Exception("not sure which target data key to use from %r" % (data_target_keys,))
+        default_data_target_key = data_target_keys[0]
+      print("Using target data key:", default_data_target_key)
     else:
-      default_data_target_key = data_target_keys[0]
-    print("Using target data key:", default_data_target_key)
+      default_data_target_key = default_data_input_key
```
Comment (albertz) [inline HDFDataset.py]: If you do that, maybe you also should add it to `data_target_keys`? Not sure if this is relevant.

File: HDFDataset.py (hunk 3 above)
Comment (albertz) [reply HDFDataset.py]: It's wrong because you changed the seqLength shape (which before was always 2 items).

Comment (ArneNx) [reply HDFDataset.py]: which I thought was the intended behaviour, but nvm. it should be back to the original version now.

Comment (albertz) [reply HDFDataset.py]: Actually I think the shape `(num_seqs, 1 + len(data_target_keys))` is more correct. When I wrote that, I thought that the HDFDataset always expected shape[-1] in {1,2}, but only later I noticed that it is also fine for multiple targets.

File: HDFDataset.py (hunk 4 above)
Comment (ArneNx) [reply HDFDataset.py]: Which change exactly are you talking about? Is this still relevant in the updated commit?

Comment (albertz) [reply HDFDataset.py]: The check that all the seq lens are the same. This was needed for the old shape of seq lengths.

Comment (albertz) [reply HDFDataset.py]: Ping?

Comment (ArneNx) [reply HDFDataset.py]: I'm sorry, I totally forgot about this. 
I don't have any time to do this properly right, but hopefully after I handed in my thesis.

Comment (albertz) [reply HDFDataset.py]: Why is that `assert` still needed? It's actually wrong, isn't it?

File: HDFDataset.py [hunk 7]
```diff
@@ -962,14 +966,17 @@ def dump_from_dataset(self, dataset, epoch=1, start_seq=0, end_seq=float("inf"),
...
+    if len(data_target_keys) > 0:
+      if len(data_target_keys) > 1:
+        if "classes" in data_target_keys:
+          default_data_target_key = "classes"
+        else:
+          raise Exception("not sure which target data key to use from %r" % (data_target_keys,))
       else:
-        raise Exception("not sure which target data key to use from %r" % (data_target_keys,))
+        default_data_target_key = data_target_keys[0]
+      print("Using target data key:", default_data_target_key)
```
Comment (albertz) [inline HDFDataset.py]: Why is the `print` only here now?
Btw, do we even need `default_data_target_key` now with the change?

**Comment by albertz** [comment]
There are conflicts now. Also, did you see my comments? Are you planning to fix this? Otherwise I will close this for now.

**Comment by albertz** [comment]
Closing this now. Please reopen if you plan to update this.
